resources/type-system/inference.md

```diff
@@ -970,167 +970,232 @@ with respect to `L` under constraints `C0`
+  to explicitly mark integer literals that have been converted, by type
+  inference, to doubles._
+
+- `@IMPLICIT_CAST<T>(m)` represents an implicit cast of the expression `m` to
+  type `T`. The runtime behavior of this construct is the same as that of `m as
+  T`, except that in the case where the cast fails, the exception thrown is a
+  `TypeError` rather than a `CastError`. The static type of
+  `@IMPLICIT_CAST<T>(m)` is `T`.
+
+- `@INT(i)` represents a literal integer with numeric value `i`. The runtime
+  behavior of this construct is to evaluate to an instance of the type `int`
+  representing `i`. The static type of `@INT(i)` is `int`. _This is used to
+  explicitly mark integer literals that have __not__ been converted, by type
+  inference, to doubles._
+
+- `@LET(T v = m_1 in m_2)` represents the operation of first evaluating `m_1`,
+  whose static type must be a subtype of `T`, storing the result in temporary
+  storage, then evaluating `m_2` in a scope in which `v` has static type `T` and
+  evaluates to the stored value. The static type of `@LET(T v = m_1 in m_2)` is
+  the same as the static type of `m_2`.
+
+  - When this specification specifies that a `@LET` expression should be created
+    using a variable `v` that does not appear in the source code, it should be
+    understood that a fresh variable is created that does not match any variable
+    that exists in the user's program. _TODO(paulberry): give an example to
+    clarify._
+
+  - _`@LET(T v = m_1 in m_2)` can be understood to be equivalent to `((v) =>
+    m_2)(m_1)`, except without the overhead of constructing and executing a
+    function closure, and without the type demotions that might normally occur
+    as a result of creating a closure._
+
+- `@PROMOTED_TYPE<T>(m)` represents an elaborated expression with static type
+  `T` whose runtime behavior is the same as that of `m`, but where it is known
+  that whenever the elaborated expression executes, the resulting value is an
+  instance of `T`. _This is used in situations where additional reasoning,
+  beyond the static type of `m`, is required to establish soundness. Wherever
+  this construct is used, the additional reasoning follows in italics. Note that
+  since `m` and `@PROMOTED_TYPE<T>(m)` have the same runtime behavior,
+  implementations can most likely elide `@PROMOTED_TYPE<T>(m)` to `m` without
+  any loss of functionality, provided they are not trying to construct a proof
+  of soundness._
 
 ## Soundness guarantees
 
-Along with the notion of _expression soundness_ described above, there are
-several other soundness guarantees made by Dart:
+An invariant of expression inference, known as _soundness_, is that when the
+elaborated form of any expression in the program is executed, it is guaranteed
+either to diverge, throw an exception, or evaluate to a value that is an
+_instance_ of its static type. _Instance_ is defined as follows: a value `v` is
```

Comment:
Interesting. @leafpetersen had a similar reaction. I pointed out to him that my interpretation was based on wanting to be consistent with the existing spec, which is what led to the non-normative comment below.

I would be totally open to switching to different terminology that is more consistent with common usage, but I would want to make the change both in this doc and in `dartLangSpec.tex` 😃

Reply:
I checked the specification. Most uses of "instance of" are about the types `bool`, `int`, `double`, `String` and `Null`. Those technically can be subclasses, but  we usually pretend they're not.

Other uses are:
> ```latex
> A function object is always an instance of some class $C$ that implements
> the class \FUNCTION{} {\ref{functionType}),
> ```

If the transitive meaning was used, it could just say "an instance that implements the class \\FUNCTION{}".

Or:
> ```latex
> If $v$ is an instance of class \code{Error} or a subclass thereof,
> ```

which is about whether throwing sets `stackTrace` on the object. 
Here it the meaning is clearly not transitive, because then it would be wrong.

And:
> ```latex
> but $e$ evaluates to an instance of a type
> which is not a subtype of \code{Iterable<\DYNAMIC>}.%
> ```
Again, this could just be "to an object which is not an instance of \\code{Iterable\<\\DYNAMIC\>}."

There are definitely also other places where "instance of" is used transitively, but I wouldn't say that the specification is consistently using it that way.

Using it transitively gets confusing, because then I don't have a good way to say that an object is an instance of a precise type. (I guess "the runtime type of the object is TheType" could work.)

I'd rather use "x is an instance of T" to mean that the runtime type of `x` is the class of `T` (and keep pretending there are no subtypes of `int` and `String`).

Reply:
> I checked the specification. Most uses of "instance of" are about the types `bool`, `int`, `double`, `String` and `Null`. Those technically can be subclasses, but we usually pretend they're not.

Ok, I think I've come around to your way of thinking. (FWIW, I was initially skeptical because the SDK files define `int`, `double`, and `String` as abstract classes, but after playing around a bunch this morning I've convinced myself that this is an implementation detail that's invisible to the user if they don't go digging into the SDK source files).

I've re-worked this PR to consistently use "instance satisfying type" instead of "instance of" in the cases where it's important to include subtypes, and I've changed this text to define "instance satisfying type" appropriately.

I might follow up with a PR to change the few places in the spec where I think we're accidentally using the "subtype" meaning of "instance of".

Reply:
After digging through the `dartLangSpec.tex`, I discovered that the terminology we use fairly consistently in the spec is "instance _implementing_ type `T`" rather than "instance _satisfying_ type `T`", so I've made a PR to switch this file to use the same convention: https://github.com/dart-lang/language/pull/3918